Cip36 lib

rooooooooob opened the pull request:
Main CIP36 work on the rust side.  This includes:
* Initial generation from `specs/cip36.cddl`
* clean-up of code (rust+WASM)
* convenience functions for converting to/from Metadata. Metadata was refactored into `core` to help with this (more so with CIP-25 as we need `chain` here anyway). (Rust only)
* crypto integration + creation of hash to sign under CIP format
* test against cip36 test vectors (weighted test doesn't pass but this seems to be due to inconsistencies with the test vectors or the CDDL spec)

rooooooooob committed:
CIP36 lib initial generation

Direct output of running `cddl-codegen` on `specs/cip36.cddl`

rooooooooob committed:
Move on-chain crypto storing + CIP36 cleanup

Moved generic types/macros for representing on-chain bytes into
`crypto/src/chain.rs` (and `crypto-wasm/src/chain.rs` for wasm).

The actual concrete types remain in the `chain` module but we needed to
store crypto on-chain for `cip36`.

rooooooooob commented on chain/Cargo.toml:
```diff
@@ -43,7 +43,7 @@ thiserror = "1.0.37"
 # wasm
 #[target.'cfg(all(target_arch = "wasm32", not(target_os = "emscripten")))'.dependencies]
-wasm-bindgen = { version = "=0.2.82", features = ["serde-serialize"] }
+wasm-bindgen = { version = "=0.2.83", features = ["serde-serialize"] }
```
This version is required for using serde wasm bindgen dependency which we added a bit ago to avoid potential cyclical dependency issues.

rooooooooob commented on crypto-wasm/src/lib.rs:
```diff
@@ -166,6 +174,12 @@ impl From<Bip32PublicKey> for core_crypto::Bip32PublicKey {
 }
 
+impl AsRef<core_crypto::Bip32PublicKey> for Bip32PublicKey {
+    fn as_ref(&self) -> &core_crypto::Bip32PublicKey {
```
I'll make `cddl-codegen` spit these out for all types since they're a lot more needed now for us. Before we could do `foo.0` to avoid cloning just to read a wasm parameter's inner rust value, but now they can be in other modules and we can't `pub` it as that messes wasm_bindgen as it wants to expose it to wasm too. This should be cleaner than doing `.0` too.

gostkin reviewed (commented):
cip36 wasm doesn't compile on my side

gostkin commented:
``` 
= note: rust-lld: error: duplicate symbol: ed25519signature_from_raw_bytes
          >>> defined in /Users/gostkin/CLionProjects/cardano-multiplatform-lib/target/wasm32-unknown-unknown/release/deps/cardano_multiplatform_lib_cip36_wasm.cardano_multiplatform_lib_cip36_wasm.62a3b95f-cgu.1.rcgu.o
          >>> defined in /Users/gostkin/CLionProjects/cardano-multiplatform-lib/target/wasm32-unknown-unknown/release/deps/libcardano_multiplatform_lib_crypto_wasm.rlib(cardano_multiplatform_lib_crypto_wasm.cardano_multiplatform_lib_crypto_wasm.9a5ad4cd-cgu.0.rcgu.o)
          
          rust-lld: error: duplicate symbol: __wbindgen_describe_ed25519signature_from_raw_bytes
          >>> defined in /Users/gostkin/CLionProjects/cardano-multiplatform-lib/target/wasm32-unknown-unknown/release/deps/cardano_multiplatform_lib_cip36_wasm.cardano_multiplatform_lib_cip36_wasm.62a3b95f-cgu.1.rcgu.o
          >>> defined in /Users/gostkin/CLionProjects/cardano-multiplatform-lib/target/wasm32-unknown-unknown/release/deps/libcardano_multiplatform_lib_crypto_wasm.rlib(cardano_multiplatform_lib_crypto_wasm.cardano_multiplatform_lib_crypto_wasm.9a5ad4cd-cgu.0.rcgu.o)
          
          rust-lld: error: duplicate symbol: __wbg_ed25519signature_free
          >>> defined in /Users/gostkin/CLionProjects/cardano-multiplatform-lib/target/wasm32-unknown-unknown/release/deps/cardano_multiplatform_lib_cip36_wasm.cardano_multiplatform_lib_cip36_wasm.62a3b95f-cgu.1.rcgu.o
          >>> defined in /Users/gostkin/CLionProjects/cardano-multiplatform-lib/target/wasm32-unknown-unknown/release/deps/libcardano_multiplatform_lib_crypto_wasm.rlib(cardano_multiplatform_lib_crypto_wasm.cardano_multiplatform_lib_crypto_wasm.9a5ad4cd-cgu.0.rcgu.o)
          

warning: `cardano-multiplatform-lib-cip36-wasm` (lib) generated 4 warnings
error: could not compile `cardano-multiplatform-lib-cip36-wasm` due to previous error; 4 warnings emitted
Error: Compiling your crate to WebAssembly failed
Caused by: failed to execute `cargo build`: exited with exit status: 101
  full command: "cargo" "build" "--lib" "--release" "--target" "wasm32-unknown-unknown"
```

rooooooooob committed:
Move Metadata into Core

Transaction Metadata now resides in core/metadata. The rest of the
metadata stays in chain but now in chain/auxdata. Now any
metadata-defined library won't need to require chain (e.g. cip25, etc)
which should greatly reduce the compiled size when crypto isn't needed.

chain/address had some refactors as I attempted to move it to its own
lib but it has too many dependencies. I kept these changes in case we
try to move it again as now it'll be easier.

Serialize's to_bytes has been reworked as discussed previously.

rooooooooob committed:
CIP36 main work (rust only)

Main post-generation work done on CIP36. This includes:
* misc clean-up and integration with address/crypto instead of raw bytes
* crypto-related stuff for creating the sign_data and hashing it
* conversion utility to/from metadata and metadata bytes
* test based on cip36 test vectors

Part of the CIP36 test vector seems to have an issue so that will need
to be resolved in the future but this should contain enough for this PR.

The WASM bindings can be fully updated after this is reviewed.

rooooooooob commented on chain/src/address.rs:
```diff
@@ -1,14 +1,40 @@
-use super::*;
 use schemars::JsonSchema;
```
These changes were mostly due to trying to move address to its own module to avoid pulling in `chain` but that isn't as easy to do as it relies on multiple things in `chain`. I kept these small changes as it will make it easier to try and move it out if we ever need to do that again to optimize space or for whatever other reason.

gostkin replied:
will the address be a separate crate in the end?

rooooooooob replied:
maybe if `chain` becomes more bloated, which could happen once we add in the other eras. The biggest contribution to both is the `crpyto` crate so the extra bit added by `chain` didn't seem significant for crates like `cip36` that needed both addresses + crypto, and addresses need hashes (in `crypto`) anyway. For flint the size differences we're talking about aren't really significant (and we'll need all crates anyway), so it might matter more for dapps that just need 1 part, but for now I don't think we need to worry about it.

rooooooooob commented on chain/src/auxdata.rs:
```diff
@@ -0,0 +1,77 @@
+/// TX auxiliary data
+use super::*;
```
This is the old `chain/src/metadata.rs` but with the metadata stuff removed and moved to `core/src/metadata.rs`

rooooooooob commented on cip36/src/lib.rs:
```diff
@@ -0,0 +1,602 @@
+        Ok(())
+    }
+
+    // these are not implementing Serialize/Deserialize as we do not keep track of the rest of the encoding metadata
```
I didn't like them implementing the same `Serialize` and having it provide `to_original_cbor_bytes()` when that isn't quite true as it might be a subset. I don't think this solution is particularly pretty but it should lead to less possible confusion. You really shouldn't be round-tripping this anyway, but instead using it for either:
1) creating the metadata in the first place
2) parsing on-chain stuff for read-only purposes that came from `Metadata` at some point (use the `Metadata` object for the round-tripping if you care about encodings)

rooooooooob commented on cip36/src/lib.rs:
```diff
@@ -0,0 +1,602 @@
+    /// # Arguments
+    ///
+    /// * `force_canonical` - Whether to encode the inner registration canonically. Should be true for hardware wallets and false otherwise.
+    pub fn hash_to_sign(&self, force_canonical: bool) -> cbor_event::Result<Vec<u8>> {
```
Do we want to have a function that combines this all in one step with a key from `crypto` to directly produce the witness? I left this as-is in case people are using it as part of other libs or need to pass it elsewhere (is this ever needed? e.g. dApp connector or whatever or would this only ever be done in one step all in the wallet?)

rooooooooob commented on core/src/lib.rs:
```diff
@@ -3,5 +3,120 @@ pub mod error;
+pub type CertificateIndex = u64;
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema, Derivative)]
+#[derivative(Eq, PartialEq, Ord="feature_allow_slow_enum", PartialOrd="feature_allow_slow_enum", Hash)]
```
pulled from `chain/src/lib.rs` since it's used in other places like metadata and seems general enough (supporting CDDL's `int` type)

rooooooooob commented on core/src/serialization.rs:
```diff
@@ -176,18 +180,21 @@ pub trait Serialize {
+    /// This gives the original bytes in the case where this was created
+    /// from bytes originally
+    fn to_original_cbor_bytes(&self) -> Vec<u8> {
         let mut buf = Serializer::new_vec();
```
having two functions seemed to be the consensus from the meetings. is this what we want to stay by? it's the most descriptive I think. I'm not sure how much of a deal it'd be space-wise for the JS lib if every single object has an extra function though.

gostkin replied:
should be fine imo

rooooooooob commented:
I'm marking this ready to review without the WASM bindings fully done as that makes more sense to finalize those after feedback on the rust code first. We can also do that as a later PR as we'll want to get this merged sooner rather than later as some of the refactoring done here will be needed to finish up CIP25.

rooooooooob commented on cip36/src/serialization.rs:
```diff
@@ -0,0 +1,550 @@
+impl Serialize for KeyRegistration {
+    fn serialize<'se, W: Write>(&self, serializer: &'se mut Serializer<W>, force_canonical: bool) -> cbor_event::Result<&'se mut Serializer<W>> {
+        serializer.write_map_sz(self.encodings.as_ref().map(|encs| encs.len_encoding).unwrap_or_default().to_len_sz(4 + if self.voting_purpose != 0 || self.encodings.as_ref().map(|encs| encs.voting_purpose_default_present).unwrap_or(false) { 1 } else { 0 }, force_canonical))?;
+        let (legacy_format, should_include_voting_purpose) = match self.delegation {
```
this bit was manually modified. I'm not sure what the desired behavior should even be. Under CDDL `.default` allows you to omit it if it's the default value. The test vectors have it present even as 0, but only in the weighted example. It shouldn't be there if it's canonical CBOR (e.g. for HW wallets I'm guessing?) so we might want an easier way (besides modifying the encodings struct manually). Should it ALWAYS be omitted in legacy examples? The CDDL spec on its own lets you do whatever, but I got the feeling that the legacy version was only for backwards compatability with CIP-15 i.e. it shouldn't be there unless you're using weighted, and even then it might be optional?

gostkin replied:
logically it looks like it's made for backwards compatibility, but it's cardano, if it turns out one day this thing can be used with legacy addresses, it can bring some trouble. @SebastienGllmt pls tell what you think when you have time

rooooooooob committed:
Merge branch 'develop' into cip36-lib

gostkin reviewed (approved):
overall lgtm, let's unblock further cip25 stuff
